### PR TITLE
WIP: Refactor

### PR DESCRIPTION
**Pull request by Eriner**
Pretty sweeping refactor, including one incomplete test due to my lack of understanding.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,54 +4,8 @@
...
-## Method Information
+##  Information
 
 Supports go versions > 1.12.1. To install add this line to your `go.mod` file 
 
 `github.com/yombu360/ANSI-378-Decoder`
-
-- **Get Record headers**
-    ```
-    GetFMDRecordHeaders(fmdByteArray []byte) (recordHeaders map[string]interface{}, err error)
```
Comment (Eriner) [inline README.md:15]: All this information can be retrieved from `godoc`.

File: record_headers.go [hunk 2]
```diff
@@ -1,62 +1,65 @@
-package ANSI378Decoder
```
Comment (Eriner) [inline record_headers.go:1]: no caps

File: record_headers.go [hunk 3]
```diff
@@ -1,62 +1,65 @@
...
+package ansi378
 
 import (
 	"encoding/binary"
-	"errors"
 )
 
-// GetFMDRecordHeaders ...
```
Comment (Eriner) [inline record_headers.go:8]: The data below should be moved to the readme, as it's general information outside the direct scope of the application code - the app doesn't care what the data values are.

File: record_headers.go [hunk 4]
```diff
@@ -1,62 +1,65 @@
...
-		recordHeaders["Equipment Compliance"] = binary.BigEndian.Uint16(fmdByteArray[14:16]) >> 12
-		recordHeaders["Equipment ID"] = binary.BigEndian.Uint16(fmdByteArray[14:16]) & 4095
-		recordHeaders["Image Size (X)"] = binary.BigEndian.Uint16(fmdByteArray[16:18])
-		recordHeaders["Image Size (Y)"] = binary.BigEndian.Uint16(fmdByteArray[18:20])
-		recordHeaders["Image Resolution (X)"] = binary.BigEndian.Uint16(fmdByteArray[20:22])
-		recordHeaders["Image Resolution (Y)"] = binary.BigEndian.Uint16(fmdByteArray[22:24])
-		recordHeaders["Views"] = uint8(fmdByteArray[24])
-		recordHeaders["Reserved"] = fmdByteArray[25]
+// RecordHeader is header data decoded from an FMD
+type RecordHeader struct {
```
Comment (Eriner) [inline record_headers.go:8]: Structs make marshaling to other formats easier, and means it's not necessary to do type declarations in random places elsewhere in the code.

File: record_headers_test.go [hunk 5]
```diff
@@ -1,48 +1,59 @@
...
-	}
-	_, err = GetFMDRecordHeaders(fmdByteArray)
-	if err != nil {
-		t.Errorf("%v", err)
-	}
+	valid6ByteFMDBase64 = "Rk1SACAyMAABLgAz/v8AAAFlAYgAxQDFAQAAAFYtQFABI3hkgH8BWHZhgFUBQnZgQEoAx3FggJIBLRpfgPwA3k1cQHYBPR1cQEsBTRxcQGAA2BdagO0A2KxaQLABDgpaQG8AkG5ZQK8AuGpZgQUBLZ5ZgMwBBmVYgPgBWJhYQM0AmwRXgMEASWVXQNYAcgVVgHQBE3lUgNEBLaBUQPkAQmBTQDYAlWlTgMABIQRTgJcBRhdTQHoAohBSgKMATwtQgK0BVwVOQKQAhQ9OgJ0BVBdNgFMAgRRLgSYAl01KgDcAdGxGQMYA9QJEQMsBYaBBgB0AoA8+gSYAj6c9QN0AHAg6gBoA+HQ6QBoAsmg3gRQAN7M2ALQBaggxAOcBbTcxAOUBXUErAMsBbU0qAAA="
+	valid6ByteFMD, _    = base64.StdEncoding.DecodeString(valid6ByteFMDBase64)
+)
 
+var recordHeaderTests = []struct {
```
Comment (Eriner) [inline record_headers_test.go:31]: Using table-driven tests mean that you can (almost) do every test based strictly by given inputs and expected outputs. Authoring multiple test functions that don't test unique code paths means the tests are far more verbose, and prone to human error when building the tests.

File: view_record.go [hunk 6]
```diff
@@ -1,55 +1,78 @@
...
 import (
 	"encoding/binary"
-	"errors"
 	"fmt"
 )
 
-// Returns the Minutiae data for each finger view
-// One Finger View Header per each finger view record. For example, 3 views will have 3  Finger View Headers
+// ViewRecords returns Minutiae data for each finger view.
+// TODO: determine what this means and clean up these docs
```
Comment (Eriner) [inline view_record.go:9]: I need help here. I don't know what any of the comments below mean.

File: view_record.go [hunk 7]
```diff
@@ -1,55 +1,78 @@
...
+	CoordinateX  uint16
+	CoordinateY  uint16
+	Reserved     uint16
+	Angle        uint8
+	Quality      uint8
+}
 
-	fingerViewRecords = make(map[string][]interface{})
+const (
+	//TODO: what is the actual minimum FMD length?
```
Comment (Eriner) [inline view_record.go:37]: Again, need help here.

File: view_record_test.go [hunk 8]
```diff
@@ -1,26 +1,23 @@
...
+package ansi378
 
-import (
-	"encoding/base64"
-	"log"
-	"testing"
-)
-
-func TestGetViewRecords_TestRecordLength(t *testing.T) {
+// The testcase below should test the validity of the ViewRecords() function in its
```
Comment (Eriner) [inline view_record_test.go:3]: This test needs to be built out more.

**Comment by vmanikes** [comment]
Thanks Matt for leaving the PR. Me and my team will review this.

File: README.md (hunk 1 above)
Comment (vmanikes) [reply README.md:15]: We are using go modules in this project, which means that GOPATH is not mandatory. But Go doc requires the project to be in go path. Last I heard is the go community is building the support for go doc. Once it releases we are more than happy to move to go doc.

File: record_headers.go (hunk 2 above)
Comment (vmanikes) [reply record_headers.go:1]: This looks good to me

File: README.md (hunk 1 above)
Comment (Eriner) [reply README.md:15]: Maybe I'm confused - I mean running `go doc`:

```
$ go doc
package ansi378 // import "github.com/flannel-dev-lab/ANSI-378-Decoder"

var ErrInvalidFMD = errors.New("fmd record is invalid")
func ViewRecords(fmd []byte, views uint8) (*[]ViewRecord, error)
type Minutiae struct{ ... }
type RecordHeader struct{ ... }
    func RecordHeaders(fmd []byte) (*RecordHeader, error)
type ViewRecord struct{ ... }
```

**Comment by vmanikes** [comment]
Hi @Eriner, I have reviewed your changes and will accept to merge onto our code, but can you create a pull request to dev branch instead of master? I will update the CONTRIBUTING.md so that new PR's will open into the dev branch

Thanks

**Comment by Eriner** [comment]
Also completely unrelated, but I sent @banjosteve an email and just want to be sure that it didn't get lost in the Spam folder. Just a ping for that.

**Pull request merged**